### PR TITLE
Introduce SqlDialect to presto-base-jdbc

### PR DESCRIPTION
**Pull request by kokosing**
Introduce SqlDialect to presto-base-jdbc

Its goal is to capture SQL dialect differences between
SQL data sources.

### CODE REVIEW

**Comment by kokosing** [comment]
My motivation behind this is that I want to have a control of how predicate is generated. 

Also I think we went to far of how allow QueryBuilder to be customized (via inheritance) and also we are using similar dialect handling in both place JdbcClient and QueryBuilder.

Both of the reasons made me think that SqlDialect is needed.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java [hunk 1]
```diff
@@ -125,30 +124,35 @@
...
     protected final ConnectionFactory connectionFactory;
-    protected final String identifierQuote;
+    protected final SqlDialect dialect;
     protected final Set<String> jdbcTypesMappedToVarchar;
     protected final boolean caseInsensitiveNameMatching;
     protected final Cache<JdbcIdentity, Map<String, String>> remoteSchemaNames;
     protected final Cache<RemoteTableNameCacheKey, Map<String, String>> remoteTableNames;
 
-    public BaseJdbcClient(BaseJdbcConfig config, String identifierQuote, ConnectionFactory connectionFactory)
+    public BaseJdbcClient(BaseJdbcConfig config, ConnectionFactory connectionFactory)
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseJdbcClient.java:133]: Keeping this is not backwards compatible, so you can remove it.

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseSqlDialect.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        implements SqlDialect
+{
+    private static final String DEFAULT_IDENTIFIER_QUOTE = "\"";
+
+    private final String identifierQuote;
+
+    public BaseSqlDialect()
+    {
+        this(DEFAULT_IDENTIFIER_QUOTE);
+    }
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseSqlDialect.java:33]: remove?

File: presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseSqlDialect.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Override
+    public String renameTable(String catalogName, String schemaName, String tableName, String newSchemaName, String newTableName)
+    {
+        return format(
+                "ALTER TABLE %s RENAME TO %s",
+                getRelation(catalogName, schemaName, tableName),
+                getRelation(catalogName, newSchemaName, newTableName));
+    }
```
Comment (findepi) [inline presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/BaseSqlDialect.java:84]: These are called in Base Jdbc which provides convenient way of customizing it.
Splitting out to dialect doesn't buy us much. There isn't going to be any direct code
reuse that I can foresee. I would keep them back there.

**Comment by findepi** [comment]
I am not sold on the change yet. Let's look at other applications solve the customizable
query generation problem. An example could be Hibernate. There, all the SQL text goes
through a dialect (things like JOIN, quotes, or LIMIT queries for instance) and dialect also
provides many "supportsXxx" methods to drive code which isn't fully delegated to dialect.
We definitely need a concept like that in Presto, and I think we already have --
the `JdbcClient` interface.

**Comment by kokosing** [comment]
I am not super fan of this either. However I see few issues that need to be solved (like using dialect handling in different classes).

Replaced with: https://github.com/prestosql/presto/pull/4636

**Pull request closed without merging**